README
PackageTools version 1.012
==========================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module does not requires other modulesor libraries.

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2007 by Ton Hospel

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension PackageTools.

1.012	2021-07-15 10:41:48 UT
	- Placeholder

1.011	2021-05-24 13:42:52 UT
	- release_pm: Use timegm with years without subtracting 1900
	- release_pm: Make bin/* and t/*.t files executable
	- release_pm: Update MANIFEST.SKIP
	- release_pm: Support README.md in Makefile.PL
	- release_pm: Add more postamble sections in Makefile.PL,
                      Remove duplicate postamble sections

1.010	2020-10-13 15:40:57 UT
        - release_pm: First write new files to a temporary before updating.
	  Untested on windows
	  (will probably fail on windows due to the typical filesystem races)
	- Added downgrade version check
	- release_pm: Added a sub_release
	- Force pdf to be "binary". They often are plain text at the start so
	  perl -B thinks they are completely plain text
	- Make -h work on windows

1.009	2012-05-03 10:32:19 UT
	- makeppd.pl: find bsdtar and zip through the PATH. Internally extend
	  the PATH with "C:/Program Files/GnuWin32/bin"
	- release_pm: Put the package name in the intial Package.pm HeadURL
	- release_pm: If we don't find any_to_blib in the path just leave the
	  string as is. If you need any_to_blib, add PackageTools to the
	  BUILD_REQUIRES in Makefile.PL

1.008	2010-12-13 12:50:46 UT
	- release_pm: Fix a very minor bug where a script name was not expanded
	  in a comment in Makefile.PL
	- any_to_blib: Add a "--define" option allowing the caller to set
	  name/value pairs
	- release_pm: Bugfix: vcs directives must end on "whitspace $" instead
	  of just "$" to be recognized
	- any_to_blib: Fix some total nonsense in the documentation
	- any_to_blib: Start by writing the processing that will be done
	  so that if it fails we know on what file it failed
	- any_to_blib: Support multiple proxy files
	- release_pm: New option "--vcs_check"
	- release_pm: On windows any_to_blib doesn't have to be "executable"

1.007	2010-10-21 13:03:02 UT
	- release_pm: Don't update Package.pm files inside a t directory
	- release_pm: Remove trailing spaces from all non-binary files
	- release_pm: Remove trailing empty lines from all non-binary files
	  (that aren't .diff or .patch files)
	- makeppd.pl: Add Carp, Errno, Exporter, File::Spec, IO::Handle, POSIX,
	  Test::Harness, Socket and URI to the known builtins
	- makeppd.pl: Remove duplicate dependecies
	- makeppd.pl: Generate provides with - instead of :: for ppm version 3
	- makeppd.pl: Fix implied dependencies on ppm version 3 to use - instead
	  of ::
	- Add ppm3 and ppm4 targets to the Makefile.PL in release_pm
	- makeppd.pl can now also try to construct a ppm3 ppd from a ppm4 ppd
	- release_pm: Add autogenerated documentation to Package.pm files
	- makeppd.pl: Added a --root option to specify the package root
	- Make cover tests work
	- makeppd.pl: ppd can have no space before final />
	- makeppd.pl: Output ppd name now comes from the package name
	- Make code slightly nicer for perlcritic
	- Generate code that is slightly nicer for perlcritic
	- release_pm: Added a "critic" target that runs perl-critic
	- Fix the magic number to get microsecond rounding correct
	- release_pm: Fix the self documentation
	- Add a few basic any_to_blib tests
	- any_to_blib: Added a "--base_dir" option to
	- any_to_blib: Removed internal "--blib" option
	  (leaving the official one)

1.006	2010-06-16 12:12:23 UT
	- makeppd.pl is ran by the ppm target instead of the $(DISTNAME).ppd
	  target. ppm_install now depends on ppm.

1.005	2010-06-07 12:38:55 UT
	- release_pm: Add support for requiring a minimum version
	- makeppd.pl: Some support for the newer ppm format. VERSION may now be
	  a float or or 4 numbers separated by ,
	- makeppd.pl: Allow "--dependency" as an alias for "--prerequisite"
	- Add some real tests
	- release_pm: Start requiring makeppd 1.014 in the generated
	  Makefile.PLs (the one with the changes in this release)
	- Added any_to_blib since it was getting more and more general use
	- release_pm: If a Package.pm file does not exist yet (or is empty) but
	  is mentioned in the MANIFEST then create a generic one
	- release_pm: Accept 0 digit versions on reread
	- makeppd.pl: Accept 0 digit versions for the package
	- Scripts find libraries relative to the executable. Mostly needed to
	  make --version work
	- release_pm: also work on deeper Makefile.PLs and t/*.t files
        - release_pm: can now generate ANY_TO_BLIB sections
	- release_pm: now recognizes a -T option in .t files
	- Use $(PERLRUN) instead of $(PERL)
	- any_to_blib: Empty file doesn't count as binary
	- any_to_blib: Empty file does not count as not ending in newline
	- release_pm: Indent autogenerated sections

1.004	2009-10-16 11:59:10 UT
	- release_pm: Ignore META.yml while versioning
	- release_pm: Fixup run suggestion for .t files
	- release_pm: Improve import_from and released handling
	- release_pm: Add md5-versions to MANIFEST if not already there
	- release_pm: Add MANIFEST.SKIP
	- release_pm: Detect duplicates in MANIFEST
	- release_pm: Force Package.pm version to be the release version
	- release_pm: now ignores more source control keywords (Id,URL,HeadURL)
	- release_pm: Source control keywords now get canonicalized differently.
	  This means that all old files that have them will count as different
	  the first time this new version is used
	- Normalize the header text in test files a bit
	- makeppd.pl: Improve error messages in  for the case of missing bsd tar
	  or commandline zip
	- makeppd.pl: Fix package mappings containing :: (they become - in the
	  ppd)
	- release_pm: Support BEGIN { $VERSION = "1.234" } style version lines
	- release_pm: Add "ppm_install" and "ppm_uninstall" make targets
	- On windows do CRLF conversion when copying to blib
	- release_pm: Add an autogenerated line to Package.pm
	- release_pm: Package.pm now uses "use vars" instead of "our". This is
	  more portable to older perls
	- release_pm: Bring $VERSION inside START_HISTORY. This means older

t/TestDrive.pm
package TestDrive;
# $Id: TestDrive.pm 5091 2012-05-15 15:09:26Z hospelt $
## no critic (UselessNoCritic MagicNumbers)
use strict;
use warnings;

our $VERSION = "1.003";

use Carp;
use FindBin qw($Bin $Script);
use Errno qw(ENOENT ESTALE);
use File::Spec;
use File::Temp qw(tempdir);
use File::Path qw(rmtree mkpath);
use File::Copy;
use Getopt::Long;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK =
    qw(ENOENT ESTALE
       $tmp_dir $t_dir $base_dir $bin_dir $me $cover $tar $zip $compress
       slurp spew rmtree mkpath cpr work_area perl_run diff);

# Allows executing programs under taint checking
delete @ENV{qw(IFS CDPATH ENV BASH_ENV)};
# Let's hope tar and gzip are in this path
$ENV{PATH} = "/usr/local/bin:/usr/bin:/bin";

$Bin =~ s{/\z}{};
if ($^O eq "MSWin32") {
    tr{\\}{/} for $Bin, $^X;
}

$Bin =~ ($^O eq "MSWin32" ?
         qr{^(((?:[A-Z]:)?(?:/[a-zA-Z0-9_:.~ -]+)*)/[a-zA-Z0-9_.-]+)/*\z} :
         qr{^(((?:/[a-zA-Z0-9_:.-]+)*)/[a-zA-Z0-9_.-]+)/*\z}) ||
    croak "Could not parse bin directory '$Bin'";
# Use untainted version lib
$Bin = $1;		## no critic (UselessNoCritic CaptureWithoutTest)
our $base_dir = $2;	## no critic (UselessNoCritic CaptureWithoutTest)

our $t_dir = $Bin;
my $t_option_file = "$t_dir/options.$^O";
our $bin_dir = "$base_dir/bin";

our $me;
if ($^O eq "MSWin32") {
    require File::Spec;
    require Win32;
    $me = Win32::LoginName();
} else {
    if (my $user = $ENV{LOGNAME}) {
        if (defined(my $uid = getpwnam($user))) {
            $me = $user if $> == $uid;
        }
    }
    $me ||= getpwuid $>;
}
die "Can't determine who I am" if !$me;
# We can basically trust $me since it came from a real system request
# Still, let's filter some weird characters
die "Unacceptable userid '$me'" if $me eq "." || $me eq "..";
$me =~ /^([0-9A-Za-z_.-]+)\z/ || die "Weird characters in userid '$me'";
# Seems ok. Untaint
$me = $1;	## no critic (UselessNoCritic CaptureWithoutTest)

# State globals
our ($cover, $tmp_dir, $tar, $zip, $compress);
our $bsd_tar	= 'bsdtar';
our $gnuwin_zip	= 'zip';

my ($keep, $leave, $strace);

sub executable {
    my ($name) = @_;
    return -x $name ? $name : undef if
        File::Spec->file_name_is_absolute($name);
    my ($volume,$directories,$file) = File::Spec->splitpath($name);
    return -x $name ? $name : undef if $volume ne "" || $directories ne "";
    for my $dir (File::Spec->path()) {
        my $try = File::Spec->catfile($dir, $tar);
        return -x $try ? $try : undef if -e $try;
    }
    return undef;
}

{
    no warnings 'recursion';
    my (@lcs_cache, @old, @new);

    # Simple recursive memoized longest common subsequence
    # This code is a bit silly. It walks the diagonal as long as the ends are
    # the same. But as soon as they differ it will suddenly fill the whole
    # remaining square. Better would be to eleminate the common start and end
    # and then fill the complete square non-recursively
    sub lcs_length {
        my ($old_l, $new_l) = @_;
        return 1 if !$old_l-- || !$new_l--;
        return 1+($lcs_cache[$old_l]{$new_l} ||= lcs_length($old_l, $new_l)) if
            $old[$old_l] eq $new[$new_l];
        my $len1 =$lcs_cache[$old_l]{$new_l+1} ||= lcs_length($old_l, $new_l+1);
        my $len2 =$lcs_cache[$old_l+1]{$new_l} ||= lcs_length($old_l+1, $new_l);
        return $len1 > $len2 ? $len1 : $len2;
    }

    sub diff {
        my $got    = shift;
        my $expect = shift;
        goto &Test::More::pass if
            defined $expect && defined $got && $expect eq $got ||
            !defined $expect && !defined $got;

        @old = $got    =~ /(.*\n|.+)/g;
        @new = $expect =~ /(.*\n|.+)/g;
        my $old_l = @old;
        my $new_l = @new;
        # Flush cache
        @lcs_cache = ();
        lcs_length($old_l, $new_l);
        # Walk the cache to recover the actual LCS
        my (@old_lcs, @new_lcs);
        while ($old_l-- && $new_l--) {
            if ($old[$old_l] eq $new[$new_l]) {
                push @old_lcs, $old_l;
                push @new_lcs, $new_l;
            } elsif ($lcs_cache[$old_l]{$new_l+1} > $lcs_cache[$old_l+1]{$new_l}) {
                $new_l++;
            } else {
                $old_l++;
            }
        }
        # Finally output
        $old_l = $new_l = 0;
        my $str = "";
        while (@old_lcs) {
            my $ol = pop @old_lcs;
            for my $i ($old_l..$ol-1) {
                $str .= "- $old[$i]";
            }
            my $nl = pop @new_lcs;
            for my $i ($new_l..$nl-1) {
                $str .= "+ $new[$i]";
            }
            $str .= "  $new[$nl]";
            $old_l = $ol+1;
            $new_l = $nl+1;
        }
        for my $i ($old_l..$#old) {
            $str .= "- $old[$i]";
        }
        for my $i ($new_l..$#new) {
            $str .= "+ $new[$i]";
        }
        $str =~ s{(\A|(?:^ .*\n){2})(?:^ .*\n)+(\z|(?:^ .*\n){2})}{$1...\n$2}mg;
        $str = "\n--- got\n+++ expected\n$str";
        $str =~ s/\n\z//;
        Test::More::diag("$str");

        goto &Test::More::fail;
    }
}

# Import a complete file and return the contents as a single string
sub slurp {
    my $file = shift;
    my ($maybe_gone, $binmode);
    if (@_) {
        if (@_ == 1) {
            $maybe_gone = shift;
        } else {
            my %params = @_;
            $maybe_gone = delete $params{maybe_gone};
            $binmode	= delete $params{binmode};
            croak "Unknown parameter ", join(", ", map "'$_'", keys %params) if
                %params;
        }
    }
    croak "filename is undefined" if !defined $file;
    open(my $fh, "<", $file) or
        $maybe_gone && ($! == ENOENT || $! == ESTALE) ?
	return undef : croak "Could not open '$file': $!";
    binmode($fh) if $binmode;
    my $rc = read($fh, my $slurp, -s $fh);
    croak "File '$file' is still growing" if $rc &&= read($fh, my $more, 1);
    croak "Error reading from '$file': $!" if !defined $rc;
    close($fh) || croak "Error while closing '$file': $!";
    return $slurp;
}

# Write remaining arguments to the file named in the first argument
# File is deleted on failure
sub spew {
    my $file = shift;
    croak "filename is undefined" if !defined $file;
    defined || croak "undef value" for @_;

    my $fsync = 0;
    my $binmode = 0;
    if (ref $_[0]) {
        ref $_[0] eq "HASH" || croak "Invalid spew parameters";
        my %params = %{+shift};

        $fsync = delete $params{fsync} if exists $params{fsync};
        $binmode = delete $params{binmode};
        croak "Unknown parameter ", join(", ", map "'$_'", keys %params) if
            %params;

    }
    open(my $fh, ">", $file) || croak "Could not create '$file': $!";
    binmode $fh if $binmode;
    eval {
        print($fh @_)	|| croak "Error writing to '$file': $!";
        $fh->flush	|| croak "Error flushing '$file': $!";
        $^O eq "MSWin32" || $fh->sync || croak "Error syncing '$file': $!" if
            $fsync;
        close($fh)	|| croak "Error closing '$file': $!";
    };
    if ($@) {
        undef $fh;
        unlink($file) || die "Could not unlink '$file' after $@";
        die $@;
    }
}

sub cpr {
    my ($from_dir, $to_dir, $dir) = @_;
    my $fd = "$from_dir/$dir";
    my $td = "$to_dir/$dir";
    if (!mkdir($td)) {
        my $err = $!;
        die "Could not mkdir($td): $err" if !-d $td;
    }
    opendir(my $dh, $fd) || die "Could not opendir '$fd': $!";
    my @files = sort readdir $dh;
    closedir($dh) || die "Error closing '$fd': $!";
    for my $f (@files) {
        next if $f eq "." || $f eq "..";
        my $file = "$fd/$f";
        lstat($file) || die "Could not lstat '$file': $!";
        if (-d _) {
            cpr($fd, $td, $f);
        } elsif (-f _) {
            File::Copy::copy($file, "$td/$f") ||
                die "Could not copy '$file' to '$td/$f': $!";
        } else {
            die "Unhandled filetype for '$file'";
        }
    }
}

sub options() {
    open(my $fh, "<", $t_option_file) ||
        die "Could not open '$t_option_file': $!";
    local $_;
    my %options;
    while (<$fh>) {
        /^\s*(\w+)\s*=\s*(.*\S)\s*\z/ ||
            die "Could not parse line $. in '$t_option_file'";
        $options{$1} = $2;
    }
    return \%options;
}

sub option($ ) {
    my ($name) = @_;
    my $options = options();
    defined $options->{$name} || croak "Unknown option $name";
    return $options->{$name};
}

sub work_area(%) {
    my (%params) = @_;

    my $copy = delete $params{copy};
    my $programs = delete $params{programs};

    croak "Unknown parameter ", join(", ", keys %params) if %params;

    local @::ARGV = @::ARGV;
    $keep	= option("KEEP");
    $strace	= option("STRACE");
    # $cover	= option("COVER");
    croak "Could not parse your command line.\n" unless
        GetOptions("leave!"		=> \$leave,
                   "keep!"		=> \$keep,
                   "cover!"		=> \$cover,
                   "strace!"		=> \$strace,
                   "help!"		=> \my $help,
                   );
    if ($help) {
        ## no critic (InputOutput::RequireCheckedSyscalls)
        print STDERR "Usage: $0 [--leave] [--keep] [--cover] [--strace] [--help]\n";
        exit 0;
    }

    if ($strace) {
        $strace =
            $^O eq "linux" ? "strace" :
            # hp: tusc
            # solaris: truss
            croak "No strace on $^O";
        $keep = 1 if !$leave;
    }
    $cover = $INC{"Devel/Cover.pm"} if !defined $cover;
    #if ($cover) {
    #    # $keep = 1;
    #    $TIMEOUT   *= 6;
    #    $AUTO_QUIT *= 6;
    #}

    my $tmp;
    if ($keep) {
        $Script =~ /^([\w-]+\.t\z)/ || die "Weird script name '$Script'";
        $keep = $1;
        # $leave = 1;
        File::Spec->catfile(File::Spec->tmpdir(), $me, "PackageTools", $keep) =~ /(.*)/s;
        $tmp = $1;
        rmtree($tmp);
        mkpath($tmp);
    } else {
        $tmp = tempdir($leave ? () : (CLEANUP => 1));
    }
    # Bring into unix form
    $tmp =~ tr{\\}{/} if $^O eq "MSWin32";

    if ($copy) {
        my $from = $copy;
        $from =~ s{/([^/]+)\z}{} || die "Cannot get dir from '$from'";
        my $dir = $1;
        cpr($from, $tmp, $dir);
    }

    if ($programs) {
        my $makefile = slurp("$base_dir/Makefile");
        ($tar) = $makefile =~ /^TAR =\s*(.*\S)\s*\n/m or
            croak "Could not get TAR from $base_dir/Makefile";
        $tar = $bsd_tar if !executable($tar);
        ($zip) = $makefile =~ /^ZIP =\s*(.*\S)\s*\n/m or
            croak "Could not get ZIP from $base_dir/Makefile";
        $zip = $gnuwin_zip unless executable($zip);
        ($compress) = $makefile =~ /^COMPRESS =\s*(.*\S)\s*\n/m or
            croak "Could not get COMPRESS from $base_dir/Makefile";
    }
    # Run is the directory for private test stuff
    mkdir ("$tmp/run") || die "Could not mkdir '$tmp/run': $!";
    $tmp_dir = $tmp;
}

sub perl_run {
    $tmp_dir || croak "Assertion: No work_area";
    open(my $old_err, ">&", "STDERR") || die "Can't dup STDERR: $!";
    open(STDERR, ">", "$tmp_dir/run/stderr") ||
        die "Could not open '$tmp_dir/run/stderr' for writing: $!";
    my $program = shift;
    # Untaint $^X
    $^X =~ ($^O eq "MSWin32" ?
             qr{^((?:[A-Z]:)?(?:/[a-zA-Z0-9_:.~ -]+)*/[a-zA-Z0-9_.-]+)\z} :
             qr{^((?:/[a-zA-Z0-9_:.-]+)*/[a-zA-Z0-9_.-]+)\z}) ||
             croak "Could not parse perl executable '$^X'";
    # No --blib since we always run the scripts from the blib directory already
    my @run = ($1, $cover ? "-MDevel::Cover" : (), $program, @_);
    # Test::More::diag("run: @run");
    my $ec = eval { system(@run) };
    my $die = $@;	# Taint failure is the only way
    open(STDERR, ">&", $old_err) || die "Can't dup old STDERR: $!";
    if ($die) {
        require Scalar::Util;
        for my $r (@run) {
            Test::More::diag("tainted $r: " .
                             (Scalar::Util::tainted($r) ? 1 : 0));
        }
        die $die;
    }
    my $err = slurp("$tmp_dir/run/stderr");
    if ($ec) {
        $err =~ s/\s+\z//;
        Test::More::fail("Unexpected exit $ec code from '@run': $err");
        exit 1;
    }
    return $err;
}

1;


t/makeppd/LogParse/blib/lib/LogParse.pm
package LogParse;
# Dummy package
our $VERSION = '1.002'; # $Revision: 3799 $
1;


t/makeppd/LogParse/blib/lib/LogParse/Info.pm
package LogParse::Info;
# Dummy package
our $VERSION = '1.000';
1;


lib/PackageTools/Package.pm
package PackageTools::Package;
# $HeadURL: xxx/lib/PackageTools/Package.pm $
# $Id: xxx $

# START HISTORY
# autogenerated by release_pm
use strict;
use warnings;
use vars qw($VERSION $SUB_VERSION $release_time %history);
$VERSION = "1.012";
$SUB_VERSION = "009";
$release_time = 1626345708;	## no critic (UselessNoCritic MagicNumbers)
%history = (
  'Changes' => {
    '1.000' => '1.000',
    '1.001' => '1.001',
    '1.002' => '1.002',
    '1.003' => '1.003',
    '1.004' => '1.004',
    '1.005' => '1.005',
    '1.006' => '1.006',
    '1.007' => '1.007',
    '1.008' => '1.008',
    '1.009' => '1.009',
    '1.010' => '1.010',
    '1.011' => '1.011',
    '1.012' => '1.012'
  },
  'MANIFEST' => {
    '1.000' => '1.000',
    '1.001' => '1.004',
    '1.002' => '1.005',
    '1.003' => '1.007',
    '1.004' => '1.010'
  },
  'MANIFEST.SKIP' => {
    '1.000' => '1.004',
    '1.001' => '1.009',
    '1.002' => '1.010',
    '1.003' => '1.012'
  },
  'Makefile.PL' => {
    '1.000' => '1.001',
    '1.001' => '1.004',
    '1.002' => '1.005',
    '1.003' => '1.006',
    '1.004' => '1.007',
    '1.005' => '1.009',
    '1.006' => '1.010',
    '1.007' => '1.012'
  },
  'README' => {
    '1.000' => '1.000',
    '1.001' => '1.001',
    '1.002' => '1.002',
    '1.003' => '1.003',
    '1.004' => '1.004',
    '1.005' => '1.005',
    '1.006' => '1.006',
    '1.007' => '1.007',
    '1.008' => '1.008',
    '1.009' => '1.009',
    '1.010' => '1.010',
    '1.011' => '1.011',
    '1.012' => '1.012'
  },
  'bin/any_to_blib' => {
    '1.000' => '1.005',
    '1.001' => '1.007',
    '1.002' => '1.008',
    '1.003' => '1.009',
    '1.004' => '1.010'
  },
  'bin/makeppd.pl' => {
    '1.011' => '1.000',
    '1.012' => '1.002',
    '1.013' => '1.004',
    '1.014' => '1.005',
    '1.015' => '1.007',
    '1.016' => '1.009',
    '1.017' => '1.010'
  },
  'bin/release_pm' => {
    '1.002' => '1.000',
    '1.003' => '1.001',
    '1.004' => '1.002',
    '1.005' => '1.003',
    '1.006' => '1.004',
    '1.007' => '1.005',
    '1.008' => '1.005',
    '1.009' => '1.006',
    '1.010' => '1.007',
    '1.011' => '1.008',
    '1.012' => '1.009',
    '1.013' => '1.010',
    '1.014' => '1.011',
    '1.015' => '1.012'
  },
  'bin/symbol_scanner' => {
    '1.000' => '1.010'
  },
  'lib/PackageTools/Package.pm' => {
    '1.000' => '1.000',
    '1.001' => '1.001',
    '1.002' => '1.002',
    '1.003' => '1.003',
    '1.004' => '1.004',
    '1.005' => '1.005',
    '1.006' => '1.006',
    '1.007' => '1.007',
    '1.008' => '1.008',
    '1.009' => '1.009',
    '1.010' => '1.010',
    '1.011' => '1.011',
    '1.012' => '1.012'
  },
  't/00_load.t' => {
    '1.000' => '1.007',
    '1.001' => '1.009'
  },
  't/00_syntax.t' => {
    '1.000' => '1.001',
    '1.001' => '1.004',
    '1.002' => '1.005'
  },
  't/01_syntax.t' => {
    '1.000' => '1.007',
    '1.001' => '1.009'
  },
  't/02_any_to_blib.t' => {
    '1.000' => '1.007',
    '1.001' => '1.009'
  },
  't/02_makeppd.t' => {
    '1.000' => '1.005',
    '1.001' => '1.006',
    '1.002' => '1.007',
    '1.003' => '1.009'
  },
  't/02_makeppd_ppm3.t' => {
    '1.000' => '1.005',
    '1.001' => '1.006',
    '1.002' => '1.007',
    '1.003' => '1.009'
  },
  't/02_makeppd_ppm3_4.t' => {
    '1.000' => '1.007',
    '1.001' => '1.009'
  },
  't/02_makeppd_ppm4_3.t' => {
    '1.000' => '1.007',
    '1.001' => '1.009'
  },
  't/TestDrive.pm' => {
    '1.000' => '1.005',
    '1.001' => '1.007',
    '1.002' => '1.009',
    '1.003' => '1.010'
  },
  't/any_to_blib/01_basic.in' => {
    '1.000' => '1.007'
  },
  't/any_to_blib/01_basic.out' => {
    '1.000' => '1.007'
  },
  't/any_to_blib/02_filter.tmpl' => {
    '1.000' => '1.007'
  },
  't/any_to_blib/02_filter.tmpl.out' => {
    '1.000' => '1.007'
  },
  't/makeppd/LogParse/LogParse.ppd' => {
    '1.000' => '1.005',
    '1.001' => '1.007'
  },
  't/makeppd/LogParse/LogParse_ppm3.ppd' => {
    '1.000' => '1.005',
    '1.001' => '1.007'
  },
  't/makeppd/LogParse/blib/lib/LogParse.pm' => {
    '1.002' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Attributes.pm' => {
    '1.000' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Info.pm' => {
    '1.000' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Info/Record.pm' => {
    '1.000' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Package.pm' => {
    '1.003' => '1.005',
    '1.005' => '1.005',
    '1.006' => '1.006'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Record.pm' => {
    '1.001' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/State.pm' => {
    '1.000' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Transaction.pm' => {
    '1.001' => '1.005'
  },
  't/makeppd/LogParse/blib/lib/LogParse/Transaction/Record.pm' => {
    '1.001' => '1.005'
  },
  't/ppm3/LogParse.ppd' => {
    '1.000' => '1.005'
  },
  'version_check' => {
    '1.000' => '1.010'
  }
);

use Carp;

my $epoch_base;

sub SUB_VERSION {
    return $SUB_VERSION;
}

sub FULL_VERSION {
    return "$VERSION.$SUB_VERSION";
}

sub release_time {
    if (!defined $epoch_base) {
	require Time::Local;
	$epoch_base = Time::Local::timegm(0,0,0,1,0,1970);	## no critic (UselessNoCritic MagicNumbers)
    }
    return $release_time + $epoch_base;
}

sub released {
    my ($package, $version) = @_;
    my $p = $package;
    $p =~ s{::}{/}g;
    my $history = $history{"lib/$p.pm"} ||
	croak "Could not find a history for package '$package'";
    my $lowest = 9**9**9;
    for my $v (keys %$history) {
	$lowest = $v if $v >= $version && $v < $lowest;
    }
    croak "No known version '$version' of package '$package'" if
	$lowest == 9**9**9;
    return $history->{$lowest};
}
1;
__END__

=for stopwords PackageTools globals

=head1 NAME

PackageTools::Package - Version and history of PackageTools

=head1 SYNOPSIS

  use PackageTools::Package;

  $epoch_time = PackageTools::Package->release_time();

  $package_version = PackageTools::Package->VERSION;

  $min_package_version = PackageTools::Package::released($module, $module_version);

=head1 DESCRIPTION

In the context of this documentation a C<package> is a set of files that
together make up a perl extension, not to be confused with the more normal
perl concept of a namespace for globals.

A package release with a certain package version number contains a number of
modules and other files with their own version numbers. This module contains a
history of which files with which versions where in which package release and
also knows when the current package was released.

This module contains a few simple methods to query this.

The version number of this Package.pm module is always equal to the version
number of the package release. This means you can use this to query this
release number and also makes it a convenient target for Makefile.PL
dependencies.

=head1 METHODS

=over

=item X<release_time>$epoch_time = PackageTools::Package->release_time()

Returns the the number of non-leap seconds since whatever time the system that
calls this function considers to be the epoch the last time
L<release_pm|release_pm(1)> was run (even if that was on another system with a
different epoch). Since the idea is to run L<release_pm|release_pm(1)> just
before releasing the package this is therefore the value of
L<time()|perlfunc/time> on the system calling this function at the moment the
package was released. This number is suitable for feeding to
L<gmtime()|perlfunc/gmtime> and L<local_time()|perlfunc/local_time>

=item X<VERSION>$package_version = PackageTools::Package->VERSION

This is the normal L<UNIVERSAL::VERSION|UNIVERSAL/VERSION> method you can use
on all modules, but this particular module is guaranteed to have a version
number that is the same as that of the package

=item X<released>$min_package_version = PackageTools::Package::released($module, $module_version)

Given a module name (e.g. Foo::Bar) and a module version number (e.g. 1.023)
returns the lowest package version in which that module (as the file
F<Foo/Bar.pm>) had at least that module version

=back

=head1 EXPORTS

None.

=head1 SEE ALSO

L<release_pm|release_pm(1)> which can be used to keep Package.pm files up to
date.

=cut

# END HISTORY
